### ISSUE TITLE
handle universes in canonicalization

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#48696]
Now that https://github.com/rust-lang/rust/pull/47861/ has landed, there is a notion of universes for type variables. For now, in order to get #48411 landed, the canonicalization code is simply using `UniverseIndex::ROOT` for the universes of type variables it creates (those universes aren't really being used anyway presently).

However, what we *ought* to be doing is recording (during canonicalization) the universe of the variables that are being fed in. We can then canonicalize those universes (what chalk calls `u_canonicalize`) and then recreate them on the other side. And finally we map the universes back.

I'll try to leave more detailed notes later. I expect to either do this or mentor this as follow-up.

cc @sgrif

**Comment by sgrif** [comment]
Definitely interested in working through this

**Comment by nikomatsakis** [comment]
@sgrif and I had a brief conversation [in gitter](https://gitter.im/rust-lang/WG-traits?at=5ab388903fe1be3704ca0dce) going into a bit more detail. I'll try to get to writing up a rustc-guide chapter here.

**Comment by nikomatsakis** [comment]
@sgrif ok so I was thinking about this last night after we talked. My PR to move skolemized things into the global tcx works, but as I said on gitter I'm now revisiting that question. Maybe you were onto something in #49598.

Perhaps we could set it up like so:

When you canonicalize, we replace skolemized regions with canonical variables, as in your PR.

We would also add information to `CanonicalVarInfo` indicating whether this was "universally" or "existentially" bound:

https://github.com/rust-lang/rust/blob/4bf76d6745046af32d0c0f33484aa15bab1c0e2c/src/librustc/infer/canonical.rs#L85-L87

Then when we instantiate each canonical variable, we would create "skolemized" variables for those things that were skolemized (in the appropriate universe):

https://github.com/rust-lang/rust/blob/4bf76d6745046af32d0c0f33484aa15bab1c0e2c/src/librustc/infer/canonical.rs#L247-L251

I think we could start with "identity mapping" between universes, in which case we would still need to carry (in the `Canonical` value) the max universe of any skolemized things we encountered, so that when we create a fresh inference context, we can make it contain that many universes. 

Once we add a real map, I think we could put the map into `CanonicalVarValues`, or maybe a new kind of struct that wraps it (`CanonicalMapping`):

https://github.com/rust-lang/rust/blob/4bf76d6745046af32d0c0f33484aa15bab1c0e2c/src/librustc/infer/canonical.rs#L76-L78

then `instantiate_query_result` would take a `CanonicalMapping` instead of just a `CanonicalVarValues`:

https://github.com/rust-lang/rust/blob/4bf76d6745046af32d0c0f33484aa15bab1c0e2c/src/librustc/infer/canonical.rs#L288

and use the info to do the mapping.
